library/lua/gui/widgets/text_area.lua

```diff
@@ -0,0 +1,184 @@
+-- Multiline text area control
+
+local Panel = require('gui.widgets.containers.panel')
+local Scrollbar = require('gui.widgets.scrollbar')
+local TextAreaContent = require('gui.widgets.text_area.text_area_content')
+local HistoryStore = require('gui.widgets.text_area.history_store')
+
+local HISTORY_ENTRY = HistoryStore.HISTORY_ENTRY
+
+TextArea = defclass(TextArea, Panel)
+
+TextArea.ATTRS{
+    init_text = '',
+    init_cursor = DEFAULT_NIL,
+    text_pen = COLOR_LIGHTCYAN,
+    ignore_keys = {'STRING_A096'},
+    select_pen = COLOR_CYAN,
+    on_text_change = DEFAULT_NIL,
+    on_cursor_change = DEFAULT_NIL,
+    one_line_mode = false,
+    debug = false
+}
+
+function TextArea:init()
+    self.render_start_line_y = 1
+
+    self.text_area = TextAreaContent{
+        frame={l=0,r=3,t=0},
+        text=self.init_text,
+
+        text_pen=self.text_pen,
+        ignore_keys=self.ignore_keys,
+        select_pen=self.select_pen,
+        debug=self.debug,
+        one_line_mode=self.one_line_mode,
+
+        on_text_change=function (text, old_text)
+            self:updateLayout()
+            if self.on_text_change then
+                self.on_text_change(text, old_text)
+            end
+        end,
+        on_cursor_change=self:callback('onCursorChange')
+    }
+    self.scrollbar = Scrollbar{
+        frame={r=0,t=1},
+        on_scroll=self:callback('onScrollbar'),
+        visible=not self.one_line_mode
+    }
+
+    self:addviews{
+        self.text_area,
+        self.scrollbar,
+    }
+    self:setFocus(true)
+end
+
+function TextArea:getText()
+    return self.text_area.text
+end
+
+function TextArea:setText(text)
+    self.text_area.history:store(
+        HISTORY_ENTRY.OTHER,
+        self:getText(),
+        self:getCursor()
+    )
+
+    return self.text_area:setText(text)
+end
+
+function TextArea:getCursor()
+    return self.text_area.cursor
+end
+
+function TextArea:setCursor(cursor_offset)
+    return self.text_area:setCursor(cursor_offset)
+end
+
+function TextArea:clearHistory()
+    return self.text_area.history:clear()
+end
+
+function TextArea:onCursorChange(cursor, old_cursor)
```

Comment:
this make code much harder to read and understand just to hide private API, I do not like it.
u propose a lot of local function that are in reality hided class function.
also, currying, which include another layer of understanding for a new dev.

maybe instead we should incorporate e.g. Python "private" functions by convention?
e.g. `TextArea:__onCursorChange`. This make code still quite simple to read and it will be very hard to have an accidental naming collision.

Reply:
I'd be ok with something like that. @lethosor do you have a preference?